File: src/main/java/solutions/binary_tree_Inorder_traversal/TreeNode.java
package solutions.binary_tree_Inorder_traversal;

/**
 * @author : Sobornov Vladimir
 * @since : 02.02.2023
 **/
public class TreeNode<T> {
    private T val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}
    TreeNode(T val) {this.val = val;}

    TreeNode(T val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    public T getValue() {
        return this.val;
    }

}

File: src/main/java/solutions/utils/TreeNode.kt
package solutions.utils

class TreeNode(private val data: Int) {
    private var left: TreeNode? = null
    private var right: TreeNode? = null

    fun find(data: Int): TreeNode? {
        if (this.data == data) {
            return this
        }

        if (this.data > data && left != null) {
            return left!!.find(data)
        }

        if (right != null) {
            return right!!.find(data)
        }
        return null
    }

    fun insert(data: Int) {
        if (data >= this.data) {
            if (this.right == null) {
                this.right = TreeNode(data)
            } else {
                insert(this.right!!.getData())
            }
        } else {
            if (this.left == null) {
                this.left = TreeNode(data)
            } else {
                insert(this.left!!.getData())
            }
        }
    }

    fun getData(): Int {
        return this.data
    }

    fun getLeftNode(): TreeNode? {
        return this.left
    }

    fun getRightNode(): TreeNode? {
        return this.right
    }

    fun setLeftNode(node: TreeNode?) {
        this.left = node
    }

    fun setRightNode(node: TreeNode?) {
        this.right = node
    }
}